_posts/2020-02-09-single-responsibility-principle.md

```diff
@@ -0,0 +1,70 @@
+3. **Data**: a profile of a person on a website, a JSON document, a message. Obviously, embodying a piece of data could only be a responsibility of a class (object), but not of a method, module, or service. A specific kind of data is **configuration**: a collection of parameters for some other method, class, or system.
+4. A piece of application's **control flow, execution, or data flow** model.  One example are methods that glue together calls to other methods:
+    ```
+    void processTransaction(Transaction t) {
+      if (isFraud(t)) { // Business logic
+        // External integration: logging
+        logger.info("Detected fraud transaction {}", t);
+        // Integration with external service
+        alertingService.sendAlert(new FraudTransactionAlert(t));
+      }
+    }
+    ```
+    Another example may be a `BufferedLogger` class which buffers logging statements in memory and manages a separate background thread that takes statements from the buffer and writes them to actual external logger:
+    ```
+    class BufferedLogger implements Logger {
+      private final Logger delegate;
+      private final ExecutorService backgroundWorker;
+      private final BlockingQueue<Statement> buffer;
+
+      BufferedLogger(Logger delegate) {
+        this.delegate = delegate;
+        this.backgroundWorker = newSingleThreadExecutor();
+        this.buffer = new ArrayBlockingQueue<>(100);
+        backgroundWorker.execute(this::writeStatementsInBackground);
+      }
+      
+      @Override public void log(Statement s) {
+        putUninterruptibly(buffer, s);
+      }
+
+      private void writeStatementsInBackground() {
+        while (true) {
+          Statement s = takeUninterruptibly(buffer);
+          delegate.log(s);
+        }
+      }
+    }
+    ```
+   `BufferedLogger` is an [active object](https://en.wikipedia.org/wiki/Active_object). Method `writeStatementsInBackground` itself has control flow responsibility.
+ 
+   In a distributed system, examples of services with this type of responsibility could be a proxy, a load balancer, or a service transparently caching responses from or buffering requests to some other service.
+
+(These types are derived from Adam Warski's classification of objects in applications which he coined in his thought-provoking post about [dependency injection in Scala](https://blog.softwaremill.com/zio-environment-meets-constructor-based-dependency-injection-6a13de6e000).)
```

Comment:
I'd move this sentence and the reference to the source up into the intro of "Types of Responsibilities" section. I was confused at first if this sentence belonged to the last example or to the whole section.

Reply:
Moved up.